### ISSUE TITLE
Write more tests

### ISSUE THREAD

**Issue by steveklabnik** [steveklabnik/rustdoc#75]
Now that we have a test framework, we should verify everything in the JSON. Right now, we only have two tests: https://github.com/steveklabnik/rustdoc/tree/master/tests/source

testing all aspects of https://github.com/steveklabnik/rustdoc/blob/master/frontend/public/data.json is good; that'd take two steps:

1. we should use `jsonapi` to validate that it's proper JSON-API
2. we should test that the proper relationships are created

**Comment by DevOrc** [comment]
For the first one were thinking of something like the [validate function](https://docs.rs/jsonapi/0.5.2/jsonapi/api/struct.JsonApiDocument.html#method.validate)? If so, I  can probably write a quick PR for this!

**Comment by steveklabnik** [comment]
@DevOrc yup! And also checking the error of [`from_str`](https://docs.rs/jsonapi/0.5.2/jsonapi/api/struct.JsonApiDocument.html#method.from_str), since that's how you'd load it up.

**Comment by mgattozzi** [comment]
1 has been completed. For those wishing to help add tests only 2 needs to be done to help close out this issue! :D

**Comment by steveklabnik** [comment]
Marking this one as easy because it fundamentally comes down to writing some `//` lines, like these ones: https://github.com/steveklabnik/rustdoc/blob/master/tests/source/modules.rs#L3

If how to do this isn't clear I'm quite happy to work with anyone on this!

**Comment by sedrik** [comment]
I started looking at writing this test but I have some questions.

```rust
#![crate_type = "lib"]

//! An example crate.
//!
//! This crate is used to test out the new rustdoc.
//!
//! It should have an example of almost everything, for testing
//! purposes.

// @has /data/type 'crate'
// @has /data/id 'nested_modules'

// @has /data/attributes/docs ' An example crate.\n\n This crate is used to test out the new rustdoc.\n\n It should have an example of almost everything, for testing\n purposes.\n'
```

This fails as the json that gets parsed is

```
{
  "data": {
    "attributes": {
      "docs": " An example crate."
    },
    "id": "nested_modules",
    "type": "crate"
  },
  "included": []
}
```

Did I write the documentation string wrong? also should the leading space be included for each line? and why does the string at `/data/attributes/docs` end with a \n (it would make sense to skip that right?).

**Comment by steveklabnik** [comment]
@sedrik that seems... odd.

>  (it would make sense to skip that right?).

We don't do any processing of the docs string on the backend, leaving that entirely to the frontend.

Related to this topic, I came across https://mattgathu.github.io/testing-rust-cli-apps/
